### PR TITLE
Refactor the Makefile and hack/ scripts

### PR DESCRIPTION
**Pull request by ConnorJC3**
**Is this a bug fix or adding new feature?**

"New Feature"/Refactoring

**What is this PR about? / Why do we need it?**

This PR contains a significant refactor of the Makefile and e2e testing scripts. Its goals are:

- **Primary Goal: Modularize `run.sh` to enable separate cluster creation, test running, and cluster deletion.** Today, running E2E tests locally requires waiting on the creation and termination of a cluster before and after the run, which takes significant time. Developers should be able to run E2E tests locally against a pre-existing cluster, as well as easily create and delete clusters for testing that follow the standard E2E setup.
- **Primary Goal: Document the available `Makefile` targets and how to use them.** New contributors should not need to go digging in the `Makefile` to discover how to perform simple tasks such as building the driver or running e2e tests.
- **Secondary Goal: Centralize dependency management.** We should have a single place that all local tools are installed - rather than the existing code which has multiple different methods of achieving the same task of installing a binary from GitHub.
- **Secondary Goal: Simply Makefile** - many of the existing makefile targets (in particular, the e2e targets) are long, unwieldy, and hard to reason about. Clean up the makefile and `hack/` directory as much as possible to remove unneeded complexity.
- **Explicit Non-Goal: Significant changes to how the image is built or how the e2e tests are ran.** I do think we should look into improvements in both of these places in the future, but this PR already contains enough changes as is. This PR should not change the behavior of our CI (other than making it run faster and easier to develop for). To this end, `eksctl.sh`, `kops.sh`, and `metrics.sh` are largely or entirely unmodified.
- **Explicit Non-Goal: Changes to Go files** - this PR is only about refactoring the testing infrastructure, not the tests themselves.

**What testing is done?** 

Significant local testing as well as CI, see the new `docs/makefile.md` for the available makefile targets to test against locally

### CODE REVIEW

**Comment by AndrewSirenko** [comment]
Still working through this PR, but I found this visualization helpful

Pre-PR: 
pre-pr-Makefile

Post-PR: 
post-pr-Makefile

Generated with [vak/makefile2dot: Visualize your Makefile using GraphViz dot utility](https://github.com/vak/makefile2dot)

**Comment by AndrewSirenko** [comment]
Can we rebase this PR to separate linting, code extraction, filename changes, and variable name changes into separate commits from the main changes to business logic? 

Perhaps commits in this order:

1. Variable name changes
2. Code extraction 
3. Filename changes
4. *Main Business logic commit* (Bonus points if there are separate commits for Each primary/secondary goal)
5. Linting

Perhaps by running `git reset master` followed by re-committing changes by hunk/File? 

Let me know if parallel ~programming~ committing would help in this endeavor. I did something similar for the original versions of #1846 and #1845 before squashing those commits. 

Also: 
> Significant local testing as well as CI, see the new docs/makefile.md for the available makefile targets to test against locally

Means that you tested each makefile target locally and on CI, correct?

**Comment by ConnorJC3** [comment]
> Can we rebase this PR to separate linting, code extraction, filename changes, and variable name changes into separate commits from the main changes to business logic?

Is there a significant advantage to doing the large amount work to split this into separate commits? They will ultimately need to be squashed for merge (as all of the commits you listed will be in a broken, partial state) and doing so greatly increases the effort required to update the PR for review comments.

> Means that you tested each makefile target locally and on CI, correct?

All the targets were tested locally, and CI runs the targets it already does (`make verify`, `make test`, and all the e2e tests)

File: docs/makefile.md
Comment (torredil) [inline docs/makefile.md:3]: np suggest the following: _The EBS CSI Driver comes with a Makefile that can be used to develop, build, test, and release the driver. This file documents the available targets ~intended for use by humans~, the parameters they support, and common usage scenarios._

**Comment by torredil** [comment]
Hitting a quota limit here:
```
2023-12-11 21:30:31 [✖]  creating OIDC provider: operation error IAM: CreateOpenIDConnectProvider, https response error StatusCode: 409, RequestID: ce25e2a1-fa71-48e4-93a9-0e26ffd23971, LimitExceeded: Cannot exceed quota for OpenIdConnectProvidersPerAccount: 100
```

cc: @dims

**Comment by dims** [comment]
@torredil cleaned up a few, please try again?

File: hack/e2e/config.sh
Comment (AndrewSirenko) [inline hack/e2e/config.sh:59]: Referencing `GINKGO` here may cause confusion to whether it is the call of ginkgo by the AWS E2E tests, versus the K8S storage conformance tests. I would recommend renaming the variable in line with the other config constants I suggested in the comments above.

(Though this is a direct upgrade already to what currently exists)

File: hack/e2e/kops/kops.sh
Comment (AndrewSirenko) [inline hack/e2e/kops/kops.sh:125]: Nit: Can you put whatever linter you used for our scripts into the makefile documentation? In a future PR we can add an update target that runs the bash linter.

**Review by AndrewSirenko — CHANGES_REQUESTED** [review]
Great job on this!! *Super* beneficial to onboarding contributors to the project, as well as maintaining maintainer sanity. Ran a few of the commands, will make sure to run through all of them when I find more time. Will start using this makefile for daily driver development and see what else I find.

File: Makefile
Comment (ConnorJC3) [reply Makefile:21]: This will be fixed when the PR is rebased

File: Makefile
Comment (ConnorJC3) [reply Makefile:115]: These are all non-default values, not sure what you mean?

File: docs/makefile.md
Comment (ConnorJC3) [reply docs/makefile.md:21]: Can you clarify the ask? Do you want me to remove the quick start guide entirely?

File: docs/makefile.md
Comment (ConnorJC3) [reply docs/makefile.md:31]: The makefile (and therefore the makefile documentation) is used by contributors, not users, and thus is biased to their needs.

File: hack/e2e/config.sh
Comment (ConnorJC3) [reply hack/e2e/config.sh:59]: We do call gingko directly in some cases, what would be your suggested names?

These are the standard names the old makefile already used.

File: hack/e2e/kops/kops.sh
Comment (ConnorJC3) [reply hack/e2e/kops/kops.sh:125]: It's already in this PR, see target `update/shfmt`

File: Makefile
Comment (ConnorJC3) [reply Makefile:259]: There used to be more targets than the directory one in a previous version of the PR. Now that there's only one I'll just move that up to the build section

File: docs/makefile.md
Comment (ConnorJC3) [reply docs/makefile.md:3]: Fixed in next version

File: docs/makefile.md
Comment (ConnorJC3) [reply docs/makefile.md:15]: Fixed in next version other than `This` -> `The`

File: Makefile
Comment (ConnorJC3) [reply Makefile:140]: That would require putting magic if statements in run.sh

File: Makefile
Comment (AndrewSirenko) [reply Makefile:113]: Ignore all of these types of comments. My brain failed to realize these are being 'passed into' the command. 😞

**Review by AndrewSirenko — APPROVED** [review]
LGTM once rebased.

**Review by torredil — APPROVED** [review]
/lgtm
/approve
/hold

**Pull request merged**
